lib/faraday/response/logger.rb

```diff
@@ -4,78 +4,93 @@
-      info('request')  { "#{env.method.upcase} #{apply_filters(env.url.to_s)}" }
-      debug('request') { apply_filters(dump_headers(env.request_headers)) } if log_headers?(:request)
-      debug('request') { apply_filters(dump_body(env[:body])) } if env[:body] && log_body?(:request)
+      @formatter.request(env)
       super
     end
 
     def on_complete(env)
-      info('response')  { "Status #{env.status}" }
-      debug('response') { apply_filters(dump_headers(env.response_headers)) } if log_headers?(:response)
-      debug('response') { apply_filters(dump_body(env[:body])) } if env[:body] && log_body?(:response)
+      @formatter.response(env)
     end
 
-    def filter(filter_word, filter_replacement)
-      @filter.push([filter_word, filter_replacement])
-    end
+    class Formatter
+      extend Forwardable
 
-    private
+      DEFAULT_OPTIONS = { headers: true, bodies: false }
 
-    def dump_headers(headers)
-      headers.map { |k, v| "#{k}: #{v.inspect}" }.join("\n")
-    end
+      def initialize(logger:, options:)
+        @logger = logger || begin
+          require 'logger'
+          ::Logger.new($stdout)
+        end
+        @filter = []
+        @options = DEFAULT_OPTIONS.merge(options)
+      end
+
+      def_delegators :@logger, :debug, :info, :warn, :error, :fatal
 
-    def dump_body(body)
-      if body.respond_to?(:to_str)
-        body.to_str
-      else
-        pretty_inspect(body)
+      def request(env)
+        info('request')  { "#{env.method.upcase} #{apply_filters(env.url.to_s)}" }
+        debug('request') { apply_filters(dump_headers(env.request_headers)) } if log_headers?(:request)
+        debug('request') { apply_filters(dump_body(env[:body])) } if env[:body] && log_body?(:request)
       end
-    end
 
-    def pretty_inspect(body)
-      require 'pp' unless body.respond_to?(:pretty_inspect)
-      body.pretty_inspect
-    end
+      def response(env)
+        info('response')  { "Status #{env.status}" }
+        debug('response') { apply_filters(dump_headers(env.response_headers)) } if log_headers?(:response)
+        debug('response') { apply_filters(dump_body(env[:body])) } if env[:body] && log_body?(:response)
+      end
 
-    def log_headers?(type)
-      case @options[:headers]
-      when Hash then @options[:headers][type]
-      else @options[:headers]
+      def filter(filter_word, filter_replacement)
+        @filter.push([filter_word, filter_replacement])
       end
-    end
 
-    def log_body?(type)
-      case @options[:bodies]
-      when Hash then @options[:bodies][type]
-      else @options[:bodies]
+      private
+
+      def dump_headers(headers)
+        headers.map { |k, v| "#{k}: #{v.inspect}" }.join("\n")
+      end
+
+      def dump_body(body)
+        if body.respond_to?(:to_str)
+          body.to_str
+        else
+          pretty_inspect(body)
+        end
+      end
+
+      def pretty_inspect(body)
+        require 'pp' unless body.respond_to?(:pretty_inspect)
```

Comment:
Any thoughts on moving this? This could be required at the top instead. The Faraday Middleware Registry should prevent this from being loaded _until_ someone uses the logger middleware for the first time.

Reply:
makes sense to me